Add Linux (gcc) support

JolifantoBambla opened the pull request:
This adds a CMake build process for Gears-Vk to make it usable on Linux.
It is currently not meant to replace the Visual Studio build process, since it doesn't replace the Post Build Helper and has significantly higher build times (it builds the dependencies instead of using pre-built binaries).
I didn't use the included DLLs for Windows builds for now, because I think it is out of scope for this pull request.
CMake builds on Windows remain experimental for now (only release builds are supported and DLLs are not moved to the correct location on Windows after the build).

I have also added a GitHub workflow which builds the `hello_world` example on Linux.
It takes quite some time, so I have only enabled it on PRs and the master branch.

My two main problems where the modified stb dependency (when building with CMake I use the original stb now) and that `AVK_USE_VMA` didn't propagate from *Gears-Vk* to executables resulting in undefined references and segfaults.
My solution to the latter is to force `AVK_USE_VMA` do be defined by the compiler, which is not optimal, but it works for now.

I haven't tested all examples since I currently don't have the hardware for it, but I've created a [template project Linux/CMake users can use to create a Gears-Vk project and start programming](https://github.com/JolifantoBambla/Gears-Vk-Starter) (it's basically a copy of the model_loader example).
It's not quite the same as the Post Build Helper, but at least its `/shaders` and `/assets` directories are copied to the same location as the executable for convenience.

Not included in this PR / Next steps:
 - fix CMake script for Windows
 - add compiler check workflow for Windows
 - use pre-built binaries where possible
 - support Clang (currently not possible, because Assimp doesn't compile with it)

JolifantoBambla commented:
@johannesugb I changed the CMake scripts now, so that the included DLLs are used on Windows. However, the DLLs are currently not copied over to the location of the built exe, so you still have to manually copy them over from the release or debug directory.
Also the shaders still don't get compiled on both platforms post build.

JolifantoBambla commented:
@johannesugb some more changes: I've added the CMake function `add_post_build_commands` which can be called on a target from a CMake script to do add post build actions to the target.
The following post build actions are added to the target:
- compile GLSL shaders from a given directory to a given directory
- create symlinks (or copy) for a list of asset file paths in a given directory
- create symlinks (or copy) DLLs (release can be enforced) to the same directory as the executable (WIndows only)

`add_post_build_commands` is already used for every example target in `CMakeLists.txt`, defaulting to creating symlinks for dependencies and using release DLLs (on Windows)

johannesugb commented:
> `AVK_USE_VMA` didn't propagate from Gears-Vk to executables resulting in undefined references and segfaults.
Hmm, I wonder why that might be. Isn't _Auto-Vk_ "linked" to _Gears-Vk_ just via including its source files? Or is _Auto-Vk_ turned into something like a static library before including it? Because if just its source files are included into _Gears-Vk_, I don't understand why the definition would not propagate.

johannesugb commented:
>  the modified stb dependency (when building with CMake I use the original stb now)

The original stb is actually fine. I just build the relevant stb files into a library via the repository [github.com/johannesugb/stb_as_library](https://github.com/johannesugb/stb_as_library) and in order to turn it into DLLs, I had to add those `__declspec`s (which I did manually, i.e. it is tedious anyways in its current state).

Would there be an easy way to turn stb into DLLs via CMake? Or maybe does such a repository exist somewhere? If so, we could use that.

Background: I'm doing that to to extend the "Always deploy release DLLs"-feature of Post Build Helper also to stb functionality. It is a bit shady, but reduces run times in Debug mode a lot (which is especially important for loading images).        

That being said, it's not worth trying to investigate into the stb-approach anyways, because the plan is to dump stb in favour of [FreeImage](https://freeimage.sourceforge.io/) according to Issue #59. 

ON THE OTHER HAND, I just see that there is some DXT support in stb, too.... but no EXR support. Oh man... we should figure out soon, which route to go in terms of image loading library.

johannesugb commented:
> support Clang (currently not possible, because Assimp doesn't compile with it)

What's the problem with Clang? Is it a known issue and will it be fixed?

JolifantoBambla commented:
> > `AVK_USE_VMA` didn't propagate from Gears-Vk to executables resulting in undefined references and segfaults.
> 
> Hmm, I wonder why that might be. Isn't _Auto-Vk_ "linked" to _Gears-Vk_ just via including its source files? Or is _Auto-Vk_ turned into something like a static library before including it? Because if just its source files are included into _Gears-Vk_, I don't understand why the definition would not propagate.

@johannesugb yes, _Auto-Vk_'s  source files are just included in _Gears-Vk_, so I also don't really know what's going on there :/
It is not _Gears-Vk_ specific though, since I encountered the same problem in my [Auto-Vk Starter template](https://github.com/JolifantoBambla/Auto-Vk-Starter), so I guess it has something to do with the order of includes in _Auto-Vk_.

JolifantoBambla commented:
> Would there be an easy way to turn stb into DLLs via CMake? Or maybe does such a repository exist somewhere? If so, we could use that.

Yes, that's possible. We would just need to add a source file including the stb files and add it as an extra target in the CMake file.
But I would need to look into the `dllexport`/`dllimport` stuff, because I've never really worked on Windows. It has something to do with the visibility of symbols right?

> ON THE OTHER HAND, I just see that there is some DXT support in stb, too.... but no EXR support. Oh man... we should figure out soon, which route to go in terms of image loading library.

Is [OpenEXR](https://github.com/AcademySoftwareFoundation/openexr) an option?

JolifantoBambla commented:
> > support Clang (currently not possible, because Assimp doesn't compile with it)
> 
> What's the problem with Clang? Is it a known issue and will it be fixed?

My bad, looks like this was only my local setup and affected only the assimp tests.
clang should work now - I've also added a job to test it.

JolifantoBambla commented:
@johannesugb more changes: if assimp is installed on linux systems it uses the installed version instead of building it from source. This also reduces the runtime of checks by half (approximately).

JolifantoBambla commented:
@johannesugb I refactored the handling of shaders such that changes to shader files are recognized and trigger a recompilation to SPIR-V. With the OS specific options (force release DLLs, symlinks, etc.) it should now be somewhat close to what the post build helper does. I documented the behavior in the *Gears-Vk* readme and [Gears-Vk-Starter readme](https://github.com/JolifantoBambla/Gears-Vk-Starter).

JolifantoBambla commented:
@johannesugb I just noticed that *Gears-Vk* didn't compile on the newest version of MSVC when using CMake as build system (I had an older version before which worked just fine), so I moved the `format_for` specializations to an extra file and included it before all other *Gears-Vk* headers in ``gvk.hpp`. see https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/bf92b0d6739f275512433dd418b8fe274ef23091

johannesugb commented on README.md:
```diff
@@ -29,6 +30,68 @@ Set up your own project:
+All shader files found in the given `glslDirectory` are added to a custom target (named `${target}_shaders`) which is added to the given `target` as a dependency.
+Shaders belonging to this custom target will be compiled before the dependent `target` is build.
+Whenever a shader file changes they will be recompiled automatically.
+Note, however, that new shader files are only added to the custom target at configure time, so if you're working with an IDE you must make sure to trigger CMake to reconfigure its build files whenever you add a new shader.
```
```suggestion
Note, however, that new shader files are only added to the custom target at configure time. This means that if you're working with an IDE, it must be ensured to trigger CMake to reconfigure its build files whenever you add new shader files.
```

I'll give you some of these "you"s ;)

johannesugb commented on README.md:
```diff
@@ -29,6 +30,68 @@ Set up your own project:
+Shaders belonging to this custom target will be compiled before the dependent `target` is build.
+Whenever a shader file changes they will be recompiled automatically.
+Note, however, that new shader files are only added to the custom target at configure time, so if you're working with an IDE you must make sure to trigger CMake to reconfigure its build files whenever you add a new shader.
+E.g. on CLion or Visual Studio you can temporarily add a `message()` call as a work-around.
```
```suggestion
As a work-around, a `message()` call can be added temporarily in CLion or Visual Studio.
```

johannesugb commented on README.md:
```diff
@@ -29,6 +30,68 @@ Set up your own project:
+If the user doesn't have the required privileges `add_post_build_commands` falls back to copying the dependencies.
+
+### Creating a New Project (CMake)
+To create a new *Gears-Vk* project using CMake you can use the [Gears-Vk-Starter template](https://github.com/JolifantoBambla/Gears-Vk-Starter).
```
```suggestion
A new *Gears-Vk* project based on CMake can conveniently be created on GitHub with the [Gears-Vk-Starter template](https://github.com/JolifantoBambla/Gears-Vk-Starter).
```

johannesugb commented on cmake/post_build_helper/compile_shaders.cmake:
```diff
@@ -0,0 +1,62 @@
+            --target-env vulkan1.2 -V "${glslDirectory}/${glslShader}" -o ${spvShaderFile})
+        message(STATUS "Compiled GLSL shader ${glslShaderFileName} to SPIR-V shader ${spvShaderFile}")
+    endforeach(glslShader)
+endfunction()
```
Is it best practice for CMake files, too, to add newlines at the end of files?

JolifantoBambla replied:
Yes, every file should end with a newline.
fixed in https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/5b7f3c20235885ba1c8a5c3be986e10cfee16a86

johannesugb commented on external/linux/include/imconfig.h:
```diff
@@ -0,0 +1,123 @@
+//-----------------------------------------------------------------------------
```
Why have these ImGui files been duplicated into the `external/linux/` folder?

JolifantoBambla replied:
See https://github.com/cg-tuwien/Gears-Vk/pull/109#commitcomment-54032014 and https://github.com/cg-tuwien/Gears-Vk/pull/109#commitcomment-54032270

johannesugb commented on framework/include/gvk.hpp:
```diff
@@ -104,6 +106,7 @@
 #include <avk/avk.hpp>
 
 // -------------------- Gears-Vk includes --------------------
+#include "format_for.hpp"
```
Okay, that's fine. I wonder what changed that made this necessary.

JolifantoBambla replied:
Yes, I'd like to know that too. I'll check the MSVC version on my old computer and investigate.

johannesugb commented on framework/include/model.hpp:
```diff
@@ -566,23 +566,23 @@ namespace gvk
 	 */
 	template <class... Types>
-	_NODISCARD constexpr std::tuple<Types&&...> additional_vertex_data(Types&&... aArgs) noexcept {
+	[[nodiscard]] constexpr std::tuple<Types&&...> additional_vertex_data(Types&&... aArgs) noexcept {
```
That's cool. Thanks.

JolifantoBambla replied:
np, it was necessary for compilers other than MSVC anyways

johannesugb commented on framework/src/context_vulkan.cpp:
```diff
@@ -674,27 +674,27 @@ namespace gvk
 			LOG_ERROR__(fmt::format("Debug Report callback with flags[{}], object-type[{}], and Message[{}]",
 				to_string(vk::DebugReportFlagsEXT{ flags }),
-				to_string(vk::DebugReportObjectTypeEXT{objectType}),
+				to_string(vk::DebugReportObjectTypeEXT(objectType)),
```
Oh.. hmm... is this something compiler-specific that only worked on MSVC?

JolifantoBambla replied:
Yes, but I'm not sure why. I'd also have written it that way.

JolifantoBambla commented on CMakeLists.txt:
```diff
@@ -0,0 +1,251 @@
+        framework/src/vk_convenience_functions.cpp
+        framework/src/window.cpp
+        framework/src/window_base.cpp
+        #framework/src/window_vulkan.cpp # is this a cg_base-only thing - remove?
```
fixed in https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/1683a424619e1c2c997552e274928888084f80c7

JolifantoBambla commented on CMakeLists.txt:
```diff
@@ -0,0 +1,251 @@
+
+# ----------------------- Options -------------------------
+
+set(gvk_AllowedLibraryTypes INTERFACE SHARED STATIC)
```
The reason I prefixed all options is to not accidentally override options in another project (a dependency or a dependent project of *Gears-Vk*).
For example, the CMake scripts of *ASSIMP* and *GLFW* both have an option called `BUILD_SHARED_LIBS` with different defaults. This is annoying when you want to have different behaviours for different libraries and I wanted to prevent such situations for future users by prefixing all options used by *Gears-Vk*.
Would you prefer a different prefix?

JolifantoBambla commented on CMakeLists.txt:
```diff
@@ -0,0 +1,251 @@
+option(gvk_ReleaseDLLsOnly "Use release DLLs for all dependencies of examples. (Windows only)" ON)
+option(gvk_CreateDependencySymlinks "Create symbolic links instead of copying dependencies of examples, i.e. DLLs (Windows only) & assets." ON)
+
+option(gvk_BuildExamples "Build all examples for Gears-Vk." OFF)
+option(gvk_BuildHelloWorld "Build example: hello_world." OFF)
+option(gvk_BuildFramebuffer "Build example: framebuffer." OFF)
+option(gvk_BuildComputeImageProcessing "Build example: compute_image_processing." OFF)
+option(gvk_BuildMultiInvokeeRendering "Build example: multi_invokee_rendering." OFF)
+option(gvk_BuildModelLoader "Build example: model_loader." OFF)
+option(gvk_BuildOrcaLoader "Build example: orca_loader." OFF)
+option(gvk_BuildRayQueryInRayTracingShaders "Build example: ray_query_in_ray_tracing_shaders." OFF)
+option(gvk_BuildRayTracingWithShadowsAndAO "Build example: ray_tracing_with_shadows_and_ao." OFF)
+option(gvk_BuildRayTracingCustomIntersection "Build example: ray_tracing_custom_intersection." OFF)
+option(gvk_BuildTextureCubemap "Build example: texture_cubemap." OFF)
+option(gvk_BuildVertexBuffers "Build example: vertex_buffers." OFF)
```
Maybe, I wasn't sure, so I added an option for every sample.
I thought that maybe people without an RTX would prefer not to build ray tracing samples.
Now that it's already there, should I remove it?

JolifantoBambla commented on README.md:
```diff
@@ -29,6 +30,68 @@ Set up your own project:
 * The steps described under section [Creating a New Project](#creating-a-new-project) might be helpful for setting up a custom Visual Studio project that links agains _Gears-Vk_.
 
+## CMake
+_Gears-Vk_ also supports building with CMake on Linux (*gcc* or *clang*) and Windows (*MSVC*).
```
fixed in https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/fecb7511821f7ff4bd1a1733d4b350681ab7fa73

johannesugb reviewed (changes requested):
## Regarding precompiled headers
> precompile gvk.hpp such that AVK_USE_VMA doesn't have to be defined by the compiler
Well... `AVK_USE_VMA` is actually not the main reason why I wanted to have precompiled headers in the first place. The compile times are the much bigger issue actually and in the Visual Studio project, it has been set-up from the beginning with precompiled headers in mind. 

Compile times, by the way, are much lower now. I.e., supporting precompiled headers is somewhat essential for Gears-Vk in its current state.

## Additional compiler flag
It doesn't compile out of the box with MSVC because of the following:
> fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
Could you add `/bigobj`  as compiler flag just to MSVC builds?

It worked for me with 
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
```
for the `project(Gears_Vk)`.

## Regarding `compile_shaders.cmake`:            
Why exactly should we not just compile every file in the `shaders/` directory? Why does it have to be tied to the file endings?     

The endings are incomplete. They lack:
- `.rcall` (callable shaders)
- `.rmiss` (miss shaders)

JolifantoBambla commented on cmake/stb.cmake:
```diff
@@ -14,8 +14,6 @@ if(UNIX)
         FetchContent_Populate(stb)
     endif()
 
-    message(STATUS "adfasdf ${stb_SOURCE_DIR}")
```
The line has been removed in the commit you've commented on - should I add that back in? ;)

JolifantoBambla commented:
> ## Regarding precompiled headers
> > precompile gvk.hpp such that AVK_USE_VMA doesn't have to be defined by the compiler
> > Well... `AVK_USE_VMA` is actually not the main reason why I wanted to have precompiled headers in the first place. The compile times are the much bigger issue actually and in the Visual Studio project, it has been set-up from the beginning with precompiled headers in mind.

Yeah, I know, but since we found out that without precompiled headers `AVK_USE_VMA` also makes the original build process fail, I thought that might be an even better reason for using them than compilation times.

> ## Additional compiler flag
> It doesn't compile out of the box with MSVC because of the following:
> 
> > fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
> > Could you add `/bigobj`  as compiler flag just to MSVC builds?
> 
> It worked for me with
> 
> ```
> set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
> ```
> 
> for the `project(Gears_Vk)`.

fixed in https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/0e4232069b78a81869df24269f2004a7a20b2d28

> ## Regarding `compile_shaders.cmake`:
> Why exactly should we not just compile every file in the `shaders/` directory? Why does it have to be tied to the file endings?

Because `glslangValidator` doesn't support arbitrary file extensions.
From the [usage message](https://github.com/KhronosGroup/glslang/blob/1978c7692f3682d2f806e8c2879be24193e9fc0f/StandAlone/StandAlone.cpp#L1736):

```
'file' can end in .<stage> for auto-stage classification, where <stage> is:
    .conf   to provide a config file that replaces the default configuration
            (see -c option below for generating a template)
    .vert   for a vertex shader
    .tesc   for a tessellation control shader
    .tese   for a tessellation evaluation shader
    .geom   for a geometry shader
    .frag   for a fragment shader
    .comp   for a compute shader
    .mesh   for a mesh shader
    .task   for a task shader
    .rgen    for a ray generation shader
    .rint    for a ray intersection shader
    .rahit   for a ray any hit shader
    .rchit   for a ray closest hit shader
    .rmiss   for a ray miss shader
    .rcall   for a ray callable shader
    .glsl   for .vert.glsl, .tesc.glsl, ..., .comp.glsl compound suffixes
    .hlsl   for .vert.hlsl, .tesc.hlsl, ..., .comp.hlsl compound suffixes
```

> The endings are incomplete. They lack:
> 
> * `.rcall` (callable shaders)
> * `.rmiss` (miss shaders)

fixed in https://github.com/cg-tuwien/Gears-Vk/pull/109/commits/805f0a666cc6fcb54e4834781d03383e1181f07d

JolifantoBambla commented:
sorry about closing it #nervoustriggerfinger

johannesugb commented on CMakeLists.txt:
```diff
@@ -0,0 +1,258 @@
+add_library(${PROJECT_NAME} ${gvk_LibraryType})
+
+# Append MSVC specific compiler flags
+if (MSVC)
+    target_compile_options(${PROJECT_NAME}
+        ${gvk_SourceScope} "/bigobj"
+        ${gvk_SourceScope} "/MP")
+endif(MSVC)
```
It looks like we need to set those compile options for _all_ projects, i.e., also the example applications. Building the library (`Gears_Vk`) works, but building `hello_world` fails with the following message:

``` 
I:\workwork\Gears-Vk\repo\auto_vk\src\avk.cpp : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
  ninja: build stopped: subcommand failed.
```

JolifantoBambla replied:
This should do the trick: 0f4b7ec

johannesugb reviewed (changes requested):
Almost there! All examples run and compile fine, except for one:

```
  [1/1] cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=examples\texture_cubemap\CMakeFiles\texture_cubemap.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\source\texture_cubemap.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\avk.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\sync.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\vk_mem_alloc.cpp.obj  /out:examples\texture_cubemap\texture_cubemap.exe /implib:examples\texture_cubemap\texture_cubemap.lib /pdb:examples\texture_cubemap\texture_cubemap.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  I:\workwork\Gears-Vk\repo\out\build\x64-Release\CMakeFiles\Gears_Vk.dir\.\cmake_pch.cxx.obj  Gears_Vk.lib  ..\..\..\external\release\lib\x64\assimp-vc140-mt.lib  ..\..\..\external\release\lib\x64\glfw3.lib  C:\VulkanSDK\1.2.182.0\Lib\vulkan-1.lib  ..\..\..\external\release\lib\x64\stb.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\post-build.bat caee8be71cd5713b"
  FAILED: examples/texture_cubemap/texture_cubemap.exe 
  cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=examples\texture_cubemap\CMakeFiles\texture_cubemap.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\source\texture_cubemap.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\avk.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\sync.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\vk_mem_alloc.cpp.obj  /out:examples\texture_cubemap\texture_cubemap.exe /implib:examples\texture_cubemap\texture_cubemap.lib /pdb:examples\texture_cubemap\texture_cubemap.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  I:\workwork\Gears-Vk\repo\out\build\x64-Release\CMakeFiles\Gears_Vk.dir\.\cmake_pch.cxx.obj  Gears_Vk.lib  ..\..\..\external\release\lib\x64\assimp-vc140-mt.lib  ..\..\..\external\release\lib\x64\glfw3.lib  C:\VulkanSDK\1.2.182.0\Lib\vulkan-1.lib  ..\..\..\external\release\lib\x64\stb.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\post-build.bat caee8be71cd5713b"
  Error copying file (if different) from "I:/workwork/Gears-Vk/repo/assets/3rd_party/yokohama_at_night" to "I:/workwork/Gears-Vk/repo/out/build/x64-Release/examples/texture_cubemap/assets/yokohama_at_night".
  Batch file failed at line 5 with errorcode 1
  ninja: build stopped: subcommand failed.
```

JolifantoBambla commented:
> Almost there! All examples compile and run fine, except for one:
> 
> ```
>   [1/1] cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=examples\texture_cubemap\CMakeFiles\texture_cubemap.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\source\texture_cubemap.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\avk.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\sync.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\vk_mem_alloc.cpp.obj  /out:examples\texture_cubemap\texture_cubemap.exe /implib:examples\texture_cubemap\texture_cubemap.lib /pdb:examples\texture_cubemap\texture_cubemap.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  I:\workwork\Gears-Vk\repo\out\build\x64-Release\CMakeFiles\Gears_Vk.dir\.\cmake_pch.cxx.obj  Gears_Vk.lib  ..\..\..\external\release\lib\x64\assimp-vc140-mt.lib  ..\..\..\external\release\lib\x64\glfw3.lib  C:\VulkanSDK\1.2.182.0\Lib\vulkan-1.lib  ..\..\..\external\release\lib\x64\stb.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\post-build.bat caee8be71cd5713b"
>   FAILED: examples/texture_cubemap/texture_cubemap.exe 
>   cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=examples\texture_cubemap\CMakeFiles\texture_cubemap.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\source\texture_cubemap.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\avk.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\sync.cpp.obj examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\__\__\auto_vk\src\vk_mem_alloc.cpp.obj  /out:examples\texture_cubemap\texture_cubemap.exe /implib:examples\texture_cubemap\texture_cubemap.lib /pdb:examples\texture_cubemap\texture_cubemap.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  I:\workwork\Gears-Vk\repo\out\build\x64-Release\CMakeFiles\Gears_Vk.dir\.\cmake_pch.cxx.obj  Gears_Vk.lib  ..\..\..\external\release\lib\x64\assimp-vc140-mt.lib  ..\..\..\external\release\lib\x64\glfw3.lib  C:\VulkanSDK\1.2.182.0\Lib\vulkan-1.lib  ..\..\..\external\release\lib\x64\stb.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && examples\texture_cubemap\CMakeFiles\texture_cubemap.dir\post-build.bat caee8be71cd5713b"
>   Error copying file (if different) from "I:/workwork/Gears-Vk/repo/assets/3rd_party/yokohama_at_night" to "I:/workwork/Gears-Vk/repo/out/build/x64-Release/examples/texture_cubemap/assets/yokohama_at_night".
>   Batch file failed at line 5 with errorcode 1
>   ninja: build stopped: subcommand failed.
> ```

fixed in 
f80a293
